romdisk/date.asm

```diff
@@ -50,6 +88,75 @@ date_main:
         WRITE()
         ret
 
+        ; Read ACSII date and fill out date part of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   YYYY-MM-DD
+        ; Does not check the '-' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the date structure, of size DATE_STRUCT_SIZE
+        ;       DE - String source.
+        ; Returns:
+        ;       The date parts of the input date struct have been filled out.
+        ;       DE - DE + 10
+        PUBLIC ascii_to_date
+ascii_to_date:
+        push hl
+        call read_bcd_pair
+        call read_bcd_pair
+        inc de ; skip '-'
+        call read_bcd_pair
+        inc de ; skip '-'
+        call read_bcd_pair
+        ; TODO calculate and fill out day-of-week
+        pop hl
+        ret
+
+        ; Read ACSII time and fill out time part of of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   HH:MM:SS
+        ; Does not check the ':' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the date structure, of size DATE_STRUCT_SIZE
+        ;       DE - String source.
+        ; Returns:
+        ;       The time parts of the input date struct have been filled out.
+        ; Modifies: BC
+        PUBLIC ascii_to_time
+ascii_to_time:
+        push hl
+        ld bc, 5
+        add hl, bc ; skip year (2 bytes), month, day, day of week
+        call read_bcd_pair
+        inc de ; skip ':'
+        call read_bcd_pair
+        inc de ; skip ':'
+        call read_bcd_pair
+        pop hl
+        ret
+
+        ; Convert an decimal ascii pair of bytes to a BCD byte. No checking.
+        ; Parameters
+        ;       DE - Pointer to first (high) ascii byte
+        ;       HL - Pointer to result byte to update
+        ; Returns:
+        ;       (HL) - 2 BCD nibbles, big endian
+        ;       HL   - HL + 1
+        ;       DE   - DE + 2
+        ; Modifies: A
+read_bcd_pair:
+        ld a, (de)
+        sub '0'
+        ld (hl),a
+        inc de
+        ld a, (de)
+        sub '0'
+        rld
+        inc de
+        inc hl
+        ret
+
 date_error:
         ; Nothing special to print before the error message
         ld de, 0
```

Comment:
In case of an error, do you think we should print a `usage` format of the command so that users are aware of how to use this command?

Reply:
I have added some checking to the date and time parsing. It checks that ascii decimal digits and separator characters are in the right place, and if not aborts with a usage message:

    usage: date [YYYY-MM-DD [HH:MM:SS]]

This doesn't pick up invalid numbers e.g. out of rage year, zero months, days > 31 etc. But it does mean if the user types e.g .`date help`, or makes a typo they will get the usage message.

I'm not too sure about the style of the format checking in `ascii_to_date` and `ascii_to_time`. Rather than aborting at the first detected error, it goes through the whole string setting A to nonzero when it detects an error. I think this is slower but clearer and more compact ? This isn't inner loop code (assuming people don't try to reset the RTC thousands of times a second) - let me know what you think.

Reply:
@chiralos I think it is okay not to check the days > 31, zero months and wrong years too, it is also the responsibility of the user to configure it properly.

For the error checking fo `ascii_to_date` and `ascii_to_time`, you can check the the returned flag after each call, that's acceptable, something like:
```
ascii_to_time:
        ld b, ':'
        call read_bcd_pair
        ret nz ; Let's suppose that read_bcd_pair returns Z flag on success 
        call check_char
        ret nz ; Let's support check_char returns Z flag on success
        call read_bcd_pair
        ret nz
        call check_char
        ret nz
        jp read_bcd_pair ; Tail call, we an jump to the last routine
``` 
Yes, we need to check every time, so it wastes some cycles on success, but it is acceptable since configuring the clock doesn't need speed.

If you really want to make this more compact (but slower), you can take the conditional call approach:
```
ascii_to_time:
        ld b, ':'
        call read_bcd_pair
        call z, check_char
        call z, read_bcd_pair
        call z, check_char
        call z, read_bcd_pair ; Tail call, we an jump to the last routine
        ret
```

Reply:
Oh nice, using the nz flag across routines as error condition cleans things up nicely.